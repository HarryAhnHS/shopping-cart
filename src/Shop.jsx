import { useEffect, useState } from "react";
import { Link } from "react-router-dom";

import Header from "./components/Header";

const Shop = () => {
    const [products, setProducts] = useState([]);
    async function getProducts() {
        const response = await fetch('https://api.escuelajs.co/api/v1/products');
        const result = await response.json();

        const resultItems = result.filter((res) => res.id > 4 && res.id < 50).map((res) => {
            return ({
                title: res.title,
                price: res.price,
                images: res.images,
                id: res.id,
            })
        })
        setProducts(resultItems);
    }
    
    useEffect(() =>{
        getProducts();
    },[])

    console.log(products)
    return (
        <>
            <Header />
            <div className="flex flex-wrap">
                {products.map((prod) => {
                    return (
                    <div className="w-1/3" key={prod.id}>
                        <h1>{prod.title}</h1>
                        <div>
                            <img src={prod.images[0]}></img>
                        </div>
                        <div>${prod.price}</div>
                    </div>
                    )
                })}
            </div>
        </>
    )
}

export default Shop;