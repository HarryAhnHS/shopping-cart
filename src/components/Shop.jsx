import { useNavigate } from "react-router-dom";
import LoadingPage from "./LoadingPage";
import ImageCarousel from "./ImageCarousel";

const Shop = ({products, isCalled, decreaseSelected, handleSelectedChange, increaseSelected, addToCart, loading}) => {

    let navigate = useNavigate(); 
    const routeChange = (id) => { 
        let path = `/shop/${id}`; 
        navigate(path);
    }
    
    console.log(isCalled)

    return (
        <>
            {loading
            ? 
                <LoadingPage />
            :
                isCalled 
                ?
                    <div className="flex flex-wrap p-6">
                        {products.map((prod) => {
                            return (
                                    <div className="box-border w-full md:w-1/2 p-3" key={prod.id}>
                                        <div className="flex rounded-lg cursor-pointer bg-[#FFFFFF] shadow-xl hover:scale-[1.03] transition-all duration-300" onClick={() => routeChange(prod.id)}>
                                            <div className="flex-none w-1/3 h-64">
                                                <ImageCarousel images={prod.images} type={'shop'}/>
                                            </div>
                                            <div className="flex-auto p-6 flex flex-col">
                                                <div className="flex flex-col">
                                                    <h1 className="flex-auto text-base md:text-lg font-semibold text-black">{prod.title}</h1>
                                                    <div className="text-base md:text-lg font-semibold text-gray-500">${prod.price}</div>
                                                </div>
                                                <div className="flex flex-1 items-center justify-center mb-5">
                                                    <button className="w-8 h-8 border" onClick={(e) => decreaseSelected(e, prod.id)}>-</button>
                                                    <input 
                                                        className="w-16 h-8 text-center" 
                                                        aria-label="selected amount"
                                                        type="number" 
                                                        min="0" 
                                                        value={prod.selected} 
                                                        onClick={(e) => e.stopPropagation()} // Stops event propagation
                                                        onChange={(e) => handleSelectedChange(e, prod.id)} 
                                                    />
                                                    <button className="w-8 h-8 border" onClick={(e) => increaseSelected(e, prod.id)}>+</button>
                                                </div>
                                                <button className="h-10 text-[0.9rem] font-semibold text-[#FFFFFF] bg-[#283618] active:scale-95 transition-all duration-100" type="submit" onClick={(e) => addToCart(e, prod.id)}>
                                                        Add to cart
                                                </button>
                                            </div>
                                        </div>
                                    </div>
                            )
                        })}
                    </div>
                :
                    <div className="flex p-60 font-bold text-lg items-center justify-center">No stock. Please try again another time</div>
            }
        </>
    )
}

export default Shop;