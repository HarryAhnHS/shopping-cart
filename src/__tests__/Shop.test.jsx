import { describe, it, expect, beforeEach } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import "@testing-library/jest-dom";
import userEvent from '@testing-library/user-event'
import { vi } from 'vitest' 
import Shop from '../components/Shop';
import { MemoryRouter, Route, Routes } from 'react-router-dom';


// Mock functions
const mockDecreaseSelected = vi.fn()
const mockHandleSelectedChange = vi.fn();
const mockIncreaseSelected = vi.fn();
const mockAddToCart = vi.fn();

const mockSingleProduct = [
    {
        id: 1,
        title: 'Product 1',
        price: 10.00,
        images: ['https://via.placeholder.com/150'],
        selected: 1
    },
]

const mockProducts = [
    {
      id: 1,
      title: 'Product 1',
      price: 10.00,
      images: ['https://via.placeholder.com/150'],
      selected: 1
    },
    {
      id: 2,
      title: 'Product 2',
      price: 20.00,
      images: ['https://via.placeholder.com/150'],
      selected: 1
    },
    {
        id: 3,
        title: 'Product 3',
        price: 250.00,
        images: ['https://via.placeholder.com/150'],
        selected: 1
    }
]

describe('Shop', () => {
    beforeEach(() => {
        // Reset mock function calls before each test
        mockDecreaseSelected.mockClear();
        mockIncreaseSelected.mockClear();
        mockAddToCart.mockClear();
    });

    it("should render shop and match snapshot", () => {
        const { container } = render(
            <MemoryRouter>
                <Shop
                    products={mockProducts}
                    isCalled={true}
                    decreaseSelected={mockDecreaseSelected}
                    handleSelectedChange={mockHandleSelectedChange}
                    increaseSelected={mockIncreaseSelected}
                    addToCart={mockAddToCart}
                    loading={false}
                />
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
    });

    it('should display the loading page when loading is true', () => {
        render(
          <MemoryRouter>
            <Shop
              products={[]}
              isCalled={false}
              decreaseSelected={mockDecreaseSelected}
              handleSelectedChange={mockHandleSelectedChange}
              increaseSelected={mockIncreaseSelected}
              addToCart={mockAddToCart}
              loading={true}
            />
          </MemoryRouter>
        );
    
        expect(screen.getByText('Loading...')).toBeInTheDocument();
    });
    
    it('should render products when isCalled is true and loading is false', () => {
        render(
          <MemoryRouter>
            <Shop
              products={mockProducts}
              isCalled={true}
              decreaseSelected={mockDecreaseSelected}
              handleSelectedChange={mockHandleSelectedChange}
              increaseSelected={mockIncreaseSelected}
              addToCart={mockAddToCart}
              loading={false}
            />
          </MemoryRouter>
        );
    
        expect(screen.getByText('Product 1')).toBeInTheDocument();
        expect(screen.getByText('Product 2')).toBeInTheDocument();
        expect(screen.getByText('Product 3')).toBeInTheDocument();
        expect(screen.getByText('$10')).toBeInTheDocument();
        expect(screen.getByText('$20')).toBeInTheDocument();
        expect(screen.getByText('$250')).toBeInTheDocument();
    });

    it('should render decrease and increase buttons', () => {
        render(
            <MemoryRouter>
            <Shop
                products={mockSingleProduct}
                isCalled={true}
                decreaseSelected={mockDecreaseSelected}
                handleSelectedChange={mockHandleSelectedChange}
                increaseSelected={mockIncreaseSelected}
                addToCart={mockAddToCart}
                loading={false}
            />
            </MemoryRouter>
        );
        
        // Check if the decrease button exists
        const decreaseButton = screen.getByRole('button', { name: '-' });
        expect(decreaseButton).toBeInTheDocument();
        
        // Check if the increase button exists
        const increaseButton = screen.getByRole('button', { name: '+' });
        expect(increaseButton).toBeInTheDocument();
    });

    it('should navigate to product detail page when a product is clicked', async () => {
        render(
            <MemoryRouter initialEntries={['/shop']}>
            <Routes>
                <Route path="/shop" element={
                <Shop
                    products={mockProducts}
                    isCalled={true}
                    decreaseSelected={mockDecreaseSelected}
                    handleSelectedChange={mockHandleSelectedChange}
                    increaseSelected={mockIncreaseSelected}
                    addToCart={mockAddToCart}
                    loading={false}
                />
                } />
                <Route path="/shop/:id" element={<div>Product Detail Page</div>} />
            </Routes>
            </MemoryRouter>
        );

        const productDiv = screen.getByText('Product 1').closest('div');
        await userEvent.click(productDiv);

        // Verify navigation
        await waitFor(() => {
            expect(screen.getByText('Product Detail Page')).toBeInTheDocument();
        });
    });

    it('should call decreaseSelected and increaseSelected with correct id when respective buttons are clicked', async () => {
        render(
            <MemoryRouter>
            <Shop
                products={mockSingleProduct}
                isCalled={true}
                decreaseSelected={mockDecreaseSelected}
                handleSelectedChange={mockHandleSelectedChange}
                increaseSelected={mockIncreaseSelected}
                addToCart={mockAddToCart}
                loading={false}
            />
            </MemoryRouter>
        );
        // Click increase button
        const increaseButton = screen.getByRole('button', { name: '+' });
        await userEvent.click(increaseButton);

        // Verify increaseSelected call
        await waitFor(() => {
            expect(mockIncreaseSelected).toHaveBeenCalledWith(expect.anything(), 1); // Check arguments
        });

        // Click decrease button
        const decreaseButton = screen.getByRole('button', { name: '-' }); // Ensure you select the correct button
        await userEvent.click(decreaseButton);

        // Verify decreaseSelected call
        await waitFor(() => {
            expect(mockDecreaseSelected).toHaveBeenCalledWith(expect.anything(), 1); // Check arguments
        }); 
    });

    it('should call addToCart with correct Id when Add to Cart button is clicked', async () => {
        render(
          <MemoryRouter>
            <Shop
              products={mockSingleProduct}
              isCalled={true}
              decreaseSelected={mockDecreaseSelected}
              handleSelectedChange={mockHandleSelectedChange}
              increaseSelected={mockIncreaseSelected}
              addToCart={mockAddToCart}
              loading={false}
            />
          </MemoryRouter>
        );
    
        const addToCartButton = screen.getByRole('button', { name: 'Add to cart' });
        await userEvent.click(addToCartButton);
        expect(mockAddToCart).toHaveBeenCalledWith(expect.any(Object), 1);
    });
});