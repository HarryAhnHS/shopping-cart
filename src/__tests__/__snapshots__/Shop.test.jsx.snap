// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop > should render shop and match snapshot 1`] = `
<div>
  <div
    class="flex flex-wrap p-6"
  >
    <div
      class="box-border w-full md:w-1/2 p-3"
    >
      <div
        class="flex rounded-lg cursor-pointer bg-[#FFFFFF] shadow-xl"
      >
        <div
          class="flex-none w-1/3 h-64"
        >
          <img
            class="inset-0 w-full h-full object-cover rounded-l-lg"
            src="https://via.placeholder.com/150"
          />
        </div>
        <div
          class="flex-auto p-6 flex flex-col"
        >
          <div
            class="flex flex-col"
          >
            <h1
              class="flex-auto text-base md:text-lg font-semibold text-black"
            >
              Product 1
            </h1>
            <div
              class="text-base md:text-lg font-semibold text-gray-500"
            >
              $
              10
            </div>
          </div>
          <div
            class="flex flex-1 items-center justify-center mb-5"
          >
            <button
              class="w-8 h-8 border"
            >
              -
            </button>
            <input
              aria-label="selected amount"
              class="w-16 h-8 text-center"
              min="0"
              type="number"
              value="1"
            />
            <button
              class="w-8 h-8 border"
            >
              +
            </button>
          </div>
          <button
            class="h-10 text-[0.9rem] font-semibold text-[#FFFFFF] bg-[#283618] active:scale-95 transition-all duration-100"
            type="submit"
          >
            Add to cart
          </button>
        </div>
      </div>
    </div>
    <div
      class="box-border w-full md:w-1/2 p-3"
    >
      <div
        class="flex rounded-lg cursor-pointer bg-[#FFFFFF] shadow-xl"
      >
        <div
          class="flex-none w-1/3 h-64"
        >
          <img
            class="inset-0 w-full h-full object-cover rounded-l-lg"
            src="https://via.placeholder.com/150"
          />
        </div>
        <div
          class="flex-auto p-6 flex flex-col"
        >
          <div
            class="flex flex-col"
          >
            <h1
              class="flex-auto text-base md:text-lg font-semibold text-black"
            >
              Product 2
            </h1>
            <div
              class="text-base md:text-lg font-semibold text-gray-500"
            >
              $
              20
            </div>
          </div>
          <div
            class="flex flex-1 items-center justify-center mb-5"
          >
            <button
              class="w-8 h-8 border"
            >
              -
            </button>
            <input
              aria-label="selected amount"
              class="w-16 h-8 text-center"
              min="0"
              type="number"
              value="1"
            />
            <button
              class="w-8 h-8 border"
            >
              +
            </button>
          </div>
          <button
            class="h-10 text-[0.9rem] font-semibold text-[#FFFFFF] bg-[#283618] active:scale-95 transition-all duration-100"
            type="submit"
          >
            Add to cart
          </button>
        </div>
      </div>
    </div>
    <div
      class="box-border w-full md:w-1/2 p-3"
    >
      <div
        class="flex rounded-lg cursor-pointer bg-[#FFFFFF] shadow-xl"
      >
        <div
          class="flex-none w-1/3 h-64"
        >
          <img
            class="inset-0 w-full h-full object-cover rounded-l-lg"
            src="https://via.placeholder.com/150"
          />
        </div>
        <div
          class="flex-auto p-6 flex flex-col"
        >
          <div
            class="flex flex-col"
          >
            <h1
              class="flex-auto text-base md:text-lg font-semibold text-black"
            >
              Product 3
            </h1>
            <div
              class="text-base md:text-lg font-semibold text-gray-500"
            >
              $
              250
            </div>
          </div>
          <div
            class="flex flex-1 items-center justify-center mb-5"
          >
            <button
              class="w-8 h-8 border"
            >
              -
            </button>
            <input
              aria-label="selected amount"
              class="w-16 h-8 text-center"
              min="0"
              type="number"
              value="1"
            />
            <button
              class="w-8 h-8 border"
            >
              +
            </button>
          </div>
          <button
            class="h-10 text-[0.9rem] font-semibold text-[#FFFFFF] bg-[#283618] active:scale-95 transition-all duration-100"
            type="submit"
          >
            Add to cart
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
